sympy/sets/sets.py

```diff
@@ -1063,16 +1064,18 @@ def _boundary(self):
         return FiniteSet(*finite_points)
 
     def _contains(self, other):
-        if not isinstance(other, Expr) or (
+        if (not isinstance(other, Expr) or
                 other is S.Infinity or
                 other is S.NegativeInfinity or
                 other is S.NaN or
-                other is S.ComplexInfinity) or other.is_extended_real is False:
+                other is S.ComplexInfinity or
+                other.is_extended_real is False or
+                other.is_real is False):
```

Comment:
It should not be necessary to check `is_real` and `is_extended_real`. The assumptions system computes implications like `is_real -> is_extended_real` and the contrapositive `! is_extended_real -> ! is_real` so these are consistent across all of `Expr` (actually all of `Basic`...).

The infinities would all be covered by the `is_real is False` check as well.

Reply:
> The assumptions system computes implications like `is_real -> is_extended_real`

Thanks for the reminder; changed. (I suppose the explicit checks for the infinities were intended as a kind of micro-optimization, but I guess the assumption system does its best, e.g. caching, to make these quick exits not really worthwhile.)

Reply:
Classes like `Infinity` have things like `is_real = False` as class attributes. The `ManagedProperties` metaclass computes all of the implications of this at class definition time so that `is_real` becomes a property that returns `False`. That means that looking up `is_real` there is probably as efficient as looking up an attribute on `S` especially if you look up multiple attributes on `S`.

We should shun using `is` checks for speed reasons and write code that is simple and clean until an actual profile/benchmark makes the benefit measurable.

Reply:
> We should shun using `is` checks for speed reasons and write code that is simple and clean until an actual profile/benchmark makes the benefit measurable

Agreed. (And thanks for the assumptions internals.)

On a side-note: I was a bit astonished that `S.NaN.is_real` evaluates to `None`. I'd have expected `False`. (I think I can see where it's coming from, but expressions that are yielding NaN are almost by definition neither real nor complex. They may have a limit, but that requires the presence of variables and such.)

Reply:
I think of nan as often representing a failed computation. Possibly calculating in a different way would have given a sensible result but we can't be sure. That means we should not really say anything definitive that it is or is not real etc.